### PR TITLE
make sure go routines watching vhosts don't go away

### PR DESCRIPTION
**Pull request by jplouis**

### CODE REVIEW

File: web/vhost.go [hunk 1]
```diff
@@ -149,6 +148,7 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
 				cancelChan := make(chan bool)
 				vregistry.vhostWatch[vhostPath] = cancelChan
 				go func(vhostID string) {
+					defer delete(vregistry.vhostWatch, vhostPath)
```
Comment (iancmcc) [inline web/vhost.go]: Thread safety? Do we need to lock the map?

File: web/vhost.go [hunk 2]
```diff
@@ -186,7 +186,17 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
...
 					}
-					vhostRegistry.WatchKey(conn, vhostID, cancelChan, processVhost, vhostWatchError)
+					// loop if error. If watch is cancelled will not return error. Blocking call
+					for {
+						err := vhostRegistry.WatchKey(conn, vhostID, cancelChan, processVhost, vhostWatchError)
+						if err == nil {
+							glog.Infof("VHostRegisty Watch %s Stopped", vhostID)
+							return
+						}
+						glog.Infof("VHostRegisty Watch %s Restarting due to %v", vhostID, err)
```
Comment (iancmcc) [inline web/vhost.go]: Typo: Registy -> Registry

File: web/vhost.go [hunk 3]
```diff
@@ -186,7 +186,17 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
 							lastChildIDs = childIDs
 						}
 					}
-					vhostRegistry.WatchKey(conn, vhostID, cancelChan, processVhost, vhostWatchError)
+					// loop if error. If watch is cancelled will not return error. Blocking call
+					for {
+						err := vhostRegistry.WatchKey(conn, vhostID, cancelChan, processVhost, vhostWatchError)
+						if err == nil {
+							glog.Infof("VHostRegisty Watch %s Stopped", vhostID)
```
Comment (iancmcc) [inline web/vhost.go]: Typo, see below

File: web/vhost.go [hunk 4]
```diff
@@ -196,17 +206,22 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
 		//cancel watching any vhosts nodes that are no longer
 		for previousVhost, cancel := range vregistry.vhostWatch {
 			if _, found := currentVhosts[previousVhost]; !found {
-				glog.V(2).Infof("Cancelling vhost watch for %s}", previousVhost)
+				glog.Infof("Cancelling vhost watch for %s}", previousVhost)
```
Comment (iancmcc) [inline web/vhost.go]: Should probably put V(2) back here.

File: web/vhost.go (hunk 2 above)
Comment (iancmcc) [inline web/vhost.go]: Also, is this logging level appropriate?

File: web/vhost.go [hunk 5]
```diff
@@ -196,17 +206,22 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
...
 				close(cancel)
 			}
 		}
 	}
-
+	cancelChan := make(chan bool)
 	for {
-		glog.V(1).Info("Running vhostRegistry.WatchRegistry")
-		vhostRegistry.WatchRegistry(poolBasedConn, cancelChan, processVhosts, vhostWatchError)
+		glog.Info("Running vhostRegistry.WatchRegistry")
```
Comment (iancmcc) [inline web/vhost.go]: V(2)

File: web/vhost.go [hunk 6]
```diff
@@ -215,7 +230,12 @@ func (sc *ServiceConfig) syncVhosts(shutdown <-chan interface{}) error {
 				close(ch)
 			}
 			return nil
-		default:
+		case err := <-watchStopped:
+			if err != nil {
+				glog.Infof("VHostRegisty Watch Restarting due to %v", err)
```
Comment (iancmcc) [inline web/vhost.go]: Typo, V(2)?

Comment (jplouis) [inline web/vhost.go]: None of the logging events happen very often. When things stop working it is hard to track down.

**Pull request merged**